metrics-core/src/main/java/com/codahale/metrics/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -0,0 +1,265 @@
+ * however it provides looser guarantees while completely avoiding locks.
+ *
+ * Looser guarantees:
+ * <ul>
+ *     <li> Updates which occur concurrently with rescaling may be discarded if the orphaned state node is updated after
+ *     rescale has replaced it. This condition has a greater probability as the rescale interval is reduced due to the
+ *     increased frequency of rescaling. {@link #rescaleThresholdNanos} values below 30 seconds are not recommended.
+ *     <li> Given a small rescale threshold, updates may attempt to rescale into a new bucket, but lose the CAS race
+ *     and update into a newer bucket than expected. In these cases the measurement weight is reduced accordingly.
+ *     <li>In the worst case, all concurrent threads updating the reservoir may attempt to rescale rather than
+ *     a single thread holding an exclusive write lock. It's expected that the configuration is set such that
+ *     rescaling is substantially less common than updating at peak load. Even so, when size is reasonably small
+ *     it can be more efficient to rescale than to park and context switch.
+ * </ul>
+ *
+ * @author <a href="mailto:ckozak@ckozak.net">Carter Kozak</a>
+ */
+public final class LockFreeExponentiallyDecayingReservoir implements Reservoir {
+
+    private static final double SECONDS_PER_NANO = .000_000_001D;
+    private static final AtomicReferenceFieldUpdater<LockFreeExponentiallyDecayingReservoir, State> stateUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(LockFreeExponentiallyDecayingReservoir.class, State.class, "state");
+
+    private final int size;
+    private final long rescaleThresholdNanos;
+    private final Clock clock;
+
+    private volatile State state;
+
+    private static final class State {
+
+        private static final AtomicIntegerFieldUpdater<State> countUpdater = AtomicIntegerFieldUpdater.newUpdater(State.class, "count");
+
+        private final double alphaNanos;
+        private final int size;
+        private final long startTick;
+        // Count is updated after samples are successfully added to the map.
+        private final ConcurrentSkipListMap<Double, WeightedSample> values;
+
+        private volatile int count;
+
+        State(
+                double alphaNanos,
+                int size,
+                long startTick,
+                int count,
+                ConcurrentSkipListMap<Double, WeightedSample> values) {
+            this.alphaNanos = alphaNanos;
+            this.size = size;
+            this.startTick = startTick;
+            this.values = values;
+            this.count = count;
+        }
+
+        private void update(long value, long timestampNanos) {
+            double itemWeight = weight(timestampNanos - startTick);
+            double priority = itemWeight / ThreadLocalRandom.current().nextDouble();
+            boolean mapIsFull = count >= size;
+            if (!mapIsFull || values.firstKey() < priority) {
```

Comment:
Two thoughts:

Maybe this should be split into two cases to avoid passing in bypassIncrement:
```java
if (count < size) {
  if (values.putIfAbsent(priority, new WeightedSample(value, itemWeight) == null) {
    countUpdater.incrementAndGet(this);
  }
} else if (values.firstKey() < priority) {
  if (values.putIfAbsent(priority, new WeightedSample(value, itemWeight) == null) {
    values.pollFirstEntry();
  }
}
```

I think there might be a race condition here though (both before and after my suggestion).
Multiple threads can read count, see that the map is not full, add the element and then increment the counter, so the map becomes too big.
Then in subsequent writes, the map will remain at that size. Perhaps that's not a problem in practice.

I am not sure exactly how that can be fixed - perhaps you could remove more than one element if the map is too big and decrement the count? But that might lead to other race conditions instead.

Maybe this would work?
```java
if (count < size) {
  if (values.putIfAbsent(priority, new WeightedSample(value, itemWeight) == null) {
    countUpdater.incrementAndGet(this);
  }
} else if (values.firstKey() < priority) {
  if (values.putIfAbsent(priority, new WeightedSample(value, itemWeight) == null) {
    values.pollFirstEntry();
  }
  while (count > size) {
    if (countUpdater.compareAndSet(this, count, count - 1)) {
      values.pollFirstEntry();
    }
  }
}
```

Reply:
Ah no, that might not work... 

> Atomically sets the field of the given object managed by this updater to the given updated value if the current value == the expected value. This method is guaranteed to be atomic with respect to other calls to compareAndSet and set, but not necessarily with respect to other changes in the field.

Reply:
I can test with two `addSample` implementations where the common path (map is already full) avoids dealing with the counter entirely. It's not clear if that would produce better performance, it's likely impacted by method size. I can try to keep `update` and `addSample*` below the 35 byte threshold to see what happens.

> Multiple threads can read count, see that the map is not full, add the element and then increment the counter, so the map becomes too big.

I think this case is expected, the `count` value may race to increment and exceed `size`, however the size of the map doesn't exceed `size` at steady state, only when new values are added prior to removing old values.
I think there is a race between the value replacement path and rescale in which rescale adds both the new and old element before an update can invoke `values.pollFirstEntry()`. Fix should be easy enough by running pollFirstEntry until the rescaled state has an acceptable count.

We could update the `count` increment to avoid exceeding `size`, but I'm not sure it would buy us anything (and it would be difficult to take advantage of weak-CAS operations for architectures like ARM where CAS is more expensive):

```java
/** Returns true if count was modified, false if the value is already equal to the maximum size. */
boolean incrementCount() {
  // hoist size avoid multiple fetches, see -XX:+TrustFinalNonStaticFields
  int max = this.size;
  while (true) {
    int countSnapshot = this.count;
    if (countSnapshot < max) {
      if (countUpdater.compareAndSet(this, countSnapshot, countSnapshot + 1)) {
        return true;
      } // else load a new countSnapshot and try again
    } else {
      return false;
    }
  }
}
```

Reply:
Since we only need to call that method for the first size calls, it would not be a big performance issue over time.

However, perhaps it's not very important to avoid that the stable map size is sometimes larger than the maximum specified size.

I am still not sure it would be a bad idea to prefill the map with dummy values to avoid this problem. Then we could get rid of the count variable altogether (the size() method is part of the interface, but is never actually called in practice, so that could be implemented with an expensive scan of the map)

Reply:
> However, perhaps it's not very important to avoid that the stable map size is sometimes larger than the maximum specified size.

when we use the increment (which can racily increment to exceed `size`), we use the result via the `countUpdater.incrementAndGet(this) > size` check to reduce the size of the map again, so the map itself doesn't have a race, there's just a chance that the `count` value may be incremented beyond `size`.

> I am still not sure it would be a bad idea to prefill the map with dummy values to avoid this problem. Then we could get rid of the count variable altogether (the size() method is part of the interface, but is never actually called in practice, so that could be implemented with an expensive scan of the map)

If there's a performance benefit, I'm happy to try it. I'm worried because there are a couple buckets of common reservoir uses:
1. Metric on a relatively hot path: In this case the reservoir is filled quickly and generally follows all the fast paths (most updates are ignored due to low priority)
2. Metric is created and never updated (feature flagged off code path)

In the second category, creating reservoir maps with 1028 (default) nodes is relatively expensive on heap. I think we would want a strong performance argument to justify the cost. What do you think?

Reply:
I think this PR is good as it is, can't see any real practical issues with it.

Creating a map with 1028 entries with different double keys but all pointing to the same value instance doesn't seem too expensive to me, but I don't mind keeping it like this. I just have a slight preference to optimize performance and code readability for the high traffic use case (where the performance is more important anyway)

Reply:
I'll generate some data so we can make an informed decision :-)

Reply:
Using the Intellij memory agent included in the idea IDE, I ran a simple main method in the debugger:
```java
    public static void main(String[] _args) {
        Reservoir reservoir = LockFreeExponentiallyDecayingReservoir.builder().build();
        System.out.println("Breakpoint 1");
        // Additional samples to avoid any potential jitter from unlikely priority collision
        for (int i = 0; i < 5000; i++) {
            reservoir.update(1L);
        }
        System.out.println("Breakpoint 2");
    }
```

At `Breakpoint 1` the reservoir has a retained size of 120 bytes.
At `Breakpoint 2` the reservoir has a retained size of 94.97 kb (Really looking forward to [valhalla](https://openjdk.java.net/projects/valhalla/) to reduce this!)

We can optimize (as you have demonstrated) using a singleton `WeightedSample` instead of creating `size` new instances, which brings us down to 63.94 kB per unused reservoir.
We can further optimize by pre-calculating boxed doubles (assuming the reservoir uses our default or fewer samples) `private static final Double[] VALUES = IntStream.range(0, 1028).mapToObj(value -> -1D - value).toArray(Double[]::new);` bringing retained heap to 36.61 kB per unused reservoir based on internal nodes used by the ConcurrentSkipListMap.

That's a significant chunk of heap! Perhaps I can measure it with a pre-release valhalla jdk later today for comparison :-)

Reply:
Good analysis! I think we can skip that for now. Perhaps I can try to make it configurable later in some way so people have a choice of using a less memory intensive version for reservoirs with very few samples or one that may be faster for fully filled reservoirs.

As I said before, I think the PR is good to go (IMO)